File: 프로그래머스/2/42842. 카펫/카펫.py

# 가로 * 세로 =  brown + yellow
# w, h 의 순서쌍 을 구하고,  w가 h보다 큰 걸 구한다.
# yellow 가로 +2 -> 전체 가로 , yellow  세로+2 ->  전체 세로
# brown = 2w+2h+4

def solution(brown, yellow):
    for height in range (1, yellow+1) :
        if yellow % height ==0 :
            width = yellow // height
    
        if 2*width + 2*height +4 == brown :
            return [width+2,height+2]
  

File: 프로그래머스/2/42842. 카펫/카펫.java
class Solution {
    public int[] solution(int brown, int yellow) {
        int[] answer = {};
        
        // brown -6 한거의 위 아래 /2 
        // yellow는 
    
        // 전체 칸수는 brown + yellow 
        // brown = h-2 * 2 + 2w = 2w +2h -4
        // yellow = (w-2) 짜리가 h-2개 있음 = w-2 * h-2 = wh -2w -2h +4
        // 2w+2h -4 + wh -2w -2h +4 = wh
        // 
        
        int s = brown + yellow;
        
        for (int h=3; h<= Math.sqrt(s); h++) {
            if (s % h !=0) {
                continue;
            }
            
            int w = s / h;
            if ((w-2) * (h-2) == yellow) {
                return new int[] {w,h};

            }
            
        }
        return new int[] {0,0};
    }
}